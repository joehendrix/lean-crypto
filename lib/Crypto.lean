import Crypto.Mceliece.Ref348864
import Crypto.Prim.AES
import Crypto.Prim.Shake
import Crypto.Random.DRBG