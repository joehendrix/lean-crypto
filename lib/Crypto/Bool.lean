namespace Bool

instance : Xor Bool := ⟨bne⟩

@[simp] theorem false_xor (a : Bool) : false ^^^ a = a := by cases a <;> rfl
@[simp] theorem xor_false (a : Bool) : a ^^^ false = a := by cases a <;> rfl
@[simp] theorem true_xor (a : Bool) : true ^^^ a = !a := by cases a <;> rfl
@[simp] theorem xor_true (a : Bool) : a ^^^ true = !a := by cases a <;> rfl

theorem xor_comm (a b : Bool) : a ^^^ b = b ^^^ a :=
  by cases a <;> simp

theorem xor_assoc (a b c : Bool) : a ^^^ b ^^^ c = a ^^^ (b ^^^ c) :=
  by cases a <;> cases b <;> simp

-- TODO(WN): Should these be in the Lean namespace? And probably Mathlib classes
-- should imply them.
instance : Lean.IsCommutative (α := Bool) HXor.hXor := ⟨xor_comm⟩

instance : Lean.IsAssociative (α := Bool) HXor.hXor := ⟨xor_assoc⟩

end Bool