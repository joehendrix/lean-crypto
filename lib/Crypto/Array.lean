import Crypto.Fin

theorem ite_same {α:Type} (p:Prop) [h:Decidable p] (a:α) : (if p then a else a) = a := by
  cases h with
  | isFalse q =>
     simp only [q]
     trivial
  | isTrue q =>
    simp only [q]
    trivial

namespace Array

theorem size_mkEmpty {α: Type _} (n:Nat) : size (mkEmpty n : Array α) = 0 := rfl

def generate (n:Nat) (f : Fin n → α) : Array α :=
  foreachFin n (λi a => a.push (f i)) (mkEmpty n)

theorem size_generate {α} (n:Nat) (f:Fin n → α) : (generate n f).size = n := by
  simp only [generate, foreachFin]
  have h : ∀(i:Nat) (p:i ≤ n) (a:Array α),
             size (foreachFin.loop n (fun i a => push a (f i)) i p a) = a.size + i
              := by
    intro i p
    induction i with
    | zero =>
      intro a
      simp only [foreachFin.loop]
      trivial
    | succ i h =>
      intro a
      simp only [foreachFin.loop, h, size_push]
      simp only [Nat.succ_add]
      trivial
  apply Eq.trans (h _ _ _)
  simp only [size_mkEmpty, Nat.zero_add]

theorem size_append {α:Type} : ∀(x y:Array α), (x ++ y).size = x.size + y.size
| a, b => by
  simp only [HAppend.hAppend, Append.append, Array.append]
  simp only [foldl, foldlM]
  simp
  have h : ∀{j k:Nat} (_ : j+k = b.size) (r:Array α) (p: b.size ≤ b.size),
            size (Id.run (foldlM.loop (fun r v => push r v) b b.size p j k r))
              = r.size + j := by
    intro j k q r p
    revert k q r
    induction j with
    | zero =>
      intro k q r
      unfold foldlM.loop
      have not_q : ¬(k < b.size) := by
        rw [q.symm]
        simp [ Nat.lt_irrefl]
      simp [not_q]
      trivial
    | succ j ind =>
      intro k q r
      unfold foldlM.loop
      have is_lt : k < b.size := by
        rw [Nat.add_comm] at q
        exact Nat.eq_add_implies_lt q
      simp [is_lt]
      apply Eq.trans
      have q_ind : j + (k + 1) = size b := by
        simp only [Nat.succ_add] at q
        exact q
      apply ind q_ind
      simp only [size_push]
      simp only [Nat.succ_add]
      trivial
  apply h
  trivial

@[simp]
theorem size_ofSubarray {α:Type} (s:Subarray α) : (ofSubarray s).size = s.stop - s.start := sorry

@[simp]
theorem start_toSubarray {α:Type} (a:Array α) (start stop : Nat) (p : start ≤ stop) (q : stop ≤ a.size)
  : (toSubarray a start stop).start = start := by simp [toSubarray, p, q]

@[simp]
theorem start_toSubarray_large {α:Type} (a:Array α) (start stop : Nat) (p : a.size ≤ start) (q : a.size ≤ stop)
  : (toSubarray a start stop).start = a.size := by
    simp only [toSubarray]
    cases Decidable.em (stop ≤ size a) with
    | inl g =>
      simp [g]
      cases Decidable.em (start ≤ stop) with
      | inl h =>
        simp [h]
        have r : start ≤ size a := Nat.le_trans h g
        exact (Nat.le_antisymm r p)
      | inr h =>
        simp [h]
        exact (Nat.le_antisymm g q)
    | inr g =>
      simp [g]
      cases Decidable.em (start ≤ a.size) with
      | inl h =>
        simp [h]
        exact (Nat.le_antisymm h p)
      | inr h =>
        simp [h]

theorem stop_toSubarray_large {α:Type} (a:Array α) (start stop : Nat) (_ : a.size ≤ start) (q : a.size ≤ stop)
  : (toSubarray a start stop).stop = a.size := by
    simp only [toSubarray]
    cases Decidable.em (stop ≤ size a) with
    | inl g =>
      simp [g]
      cases Decidable.em (start ≤ stop) with
      | inl h =>
        simp [h]
        exact (Nat.le_antisymm g q)
      | inr h =>
        simp [h]
        exact (Nat.le_antisymm g q)
    | inr g =>
      simp [g]
      cases Decidable.em (start ≤ a.size) with
      | inl h => simp [h]
      | inr h => simp [h]

@[simp]
theorem stop_toSubarray {α:Type} (a:Array α) (start stop : Nat) (p : start ≤ stop) (q : stop ≤ a.size)
  : (toSubarray a start stop).stop = stop := by simp [toSubarray, p, q]

theorem size_empty {α:Type _} : Array.size (#[] : Array α) = 0 := by rfl

@[simp]
theorem size_extract {α:Type} (src: Array α) (start stop : Nat)
: (src.extract start stop).size = min stop src.size - start := by
  simp [extract, toSubarray]
  cases Decidable.em (stop ≤ size src) with
  | inl g =>
    simp [g, min]
    cases Decidable.em (start ≤ stop) with
    | inl h => simp [h]
    | inr h =>
      have q := Nat.gt_of_not_le h
      simp [h, Nat.sub_self, Nat.lt_implies_zero_sub q]
  | inr g =>
    simp [g, min]
    cases Decidable.em (start ≤ size src) with
    | inl h =>
      simp [h]
    | inr h =>
      have q := Nat.gt_of_not_le h
      simp [h, Nat.sub_self, Nat.lt_implies_zero_sub q]

theorem extract_all (a:Array α)
  : (a.extract 0 a.size) = a := by
    admit

theorem extract_end_empty {a:Array α} {i : Nat} (p : i ≥ a.size) (j : Nat)
  :(a.extract i j) = Array.empty := by
    admit


theorem append_empty {α} (a:Array α) : a ++ empty = a := by
  admit

theorem append_data {α} (a b:Array α) : (a ++ b).data = a.data ++ b.data := by
  admit

theorem size_qsort {α} [Inhabited α] (a:Array α) (lt : α → α → Bool)
  : Array.size (Array.qsort a lt) = a.size := by
  admit

@[simp]
theorem size_setD (a:Array α) : (a.setD i e).size = a.size := by
  simp only [setD]
  cases Decidable.em (i < a.size) with
    | inl h => simp [h]
    | inr h => simp [h]

theorem size_set! (a:Array α) : (a.set! i e).size = a.size := by
  simp [set!, size_setD]

end Array
