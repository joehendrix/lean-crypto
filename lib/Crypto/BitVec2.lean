import Crypto.Range
import Crypto.UInt8
import Crypto.ToMathlib

import Smt.Data.BitVec

namespace BitVec

def lsb_getAux (x : BitVec m) (i : Nat) : Bool :=
  (x.val &&& (1 <<< i)) ≠ 0

@[implementedBy lsb_getAux]
def lsb_get! (x : BitVec m) (i : Nat) : Bool :=
  BitVec.lsbGet x i

def lsb_set! (x : BitVec m) (i : Nat) (c : Bool) : BitVec m :=
  if c then
    x ||| ⟨1 <<< i, sorry⟩
  else
    x &&& ⟨((1 <<< m) - 1 - (1 <<< i)), sorry⟩

/--
Update index to use most-significant bytes, but least-significant bit
ordering within bytes.

This may be removed once compatibility with C is not needed.
-/
def msbb_fix (m : Nat) (i : Nat) : Nat :=
  let j := (m-1)-i
  -- Reverse bit order within bytes (see if we can fix this)
  ((j >>> 3) <<< 3) ||| (0x7 - (j &&& 0x7))

def msbb_get! (x : BitVec m) (i : Nat) : Bool := x.lsb_get! (msbb_fix m i)

def msbb_set! (x : BitVec m) (i : Nat) (c : Bool) : BitVec m :=
  x.lsb_set! (msbb_fix m i) c

protected def toBinary (x : BitVec n) : String :=
  let l := Nat.toDigits 2 x.val
  "0b" ++ String.mk (List.replicate (n - l.length) '0' ++ l)

protected def toHex (x : BitVec n) : String :=
  let l := Nat.toDigits 16 x.val
  "0x" ++ String.mk (List.replicate (n/4 - l.length) '0' ++ l)

protected def toHex2 (x : BitVec n) : String := Id.run do
  let mut s : String := "0x"
  for i in range 0 (n/8) do
    let b := UInt8.ofNat (x.val >>> (8*i))
    s := s ++ b.toHex
  pure s

instance : ToString (BitVec n) := ⟨BitVec.toHex2⟩

def reverse (x : BitVec n) : BitVec n := Id.run do
  let mut r : Nat := 0
  for i in range 0 n do
    r := r <<< 1
    if x.lsb_get! i then
      r := r + 1
  pure ⟨r, sorry⟩

protected def foldl (f : α → Bool → α) (x : BitVec n) (a : α) : α := Id.run do
  let mut r := a
  for i in range 0 n do
    r := f r (x.msbb_get! i)
  pure r

protected def take_lsb (x : BitVec m) (n : Nat) : BitVec n :=
  ⟨x.val &&& 1 <<< n - 1, sorry⟩

protected def take_msb (x : BitVec m) (n : Nat) : BitVec n :=
  ⟨x.val >>> (m-n), sorry⟩

theorem eq_of_val_eq : ∀ {x y : BitVec n}, x.val = y.val → x = y
  | ⟨_,_⟩, _, rfl => rfl

theorem ne_of_val_ne {x y : BitVec n} (h : x.val ≠ y.val) : x ≠ y :=
  λ h' => absurd (h' ▸ rfl) h

protected def decideEq : (x y : BitVec n) → Decidable (x = y) :=
  fun ⟨i, _⟩ ⟨j, _⟩ =>
    match decEq i j with
    | isTrue h  => isTrue (eq_of_val_eq h)
    | isFalse h => isFalse (ne_of_val_ne h)

instance : DecidableEq (BitVec n) := BitVec.decideEq

protected def generate_lsb (n : Nat) (f : Fin n → Bool) : BitVec n := Id.run do
  let mut r : Nat := 0

  for i in range 0 n do
    let b := f ⟨i, sorry⟩
    r := (if b then 1 <<< i else 0) ||| r

  ⟨r, sorry⟩

protected def generate_msbb (n : Nat) (f : Fin n → Bool) : BitVec n := Id.run do
  let mut r : Nat := 0

  let m := n % 8
  if m ≠ 0 then
    let mut w : Nat := 0
    for j in range 0 m do
      let b := f ⟨j, sorry⟩
      w := if b then 1 <<< j ||| w else w
    r := w

  for i in range 0 (n/8) do
    let mut w : Nat := 0
    for j in range 0 8 do
      let b := f ⟨m+8*i+j, sorry⟩
      w := if b then 1 <<< j ||| w else w
    r := r <<< 8 ||| w

  ⟨r, sorry⟩

end BitVec
