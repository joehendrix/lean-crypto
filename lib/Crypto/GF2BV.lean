import Crypto.ToMathlib
import Crypto.BitVec
import Crypto.GF2Poly

import Smt.Tactic.WHNFSmt

/-! Concrete implementations of GF(2)[X] and GF(2^k) operations as bitvector circuits. -/

namespace GF2BVPoly

/-- Semantic interpretation of a bitvector as a polynomial. -/
def interp (x : BitVec w) : GF2Poly :=
  GF2Poly.mk x.val

-- TODO(WN): might have to define in terms of BV operations
def polyDegree (x : BitVec w) : Nat :=
  (interp x).degree

def polyAdd (a b : BitVec w) : BitVec w := a ^^^ b

/-- Long multiplication in GF(2)[x] translated from Cryptol reference.

```cryptol
pmult : {u, v} (fin u, fin v) => [1 + u] -> [1 + v] -> [1 + u + v]
pmult x y = last zs
  where
    zs = [0] # [ (z << 1) ^ (if yi then 0 # x else 0) | yi <- y | z <- zs ]
``` -/
def polyMul (a : BitVec w) (b : BitVec v) : BitVec (w+v) := Id.run do
  let wOut := w + v
  let_opaque a := a.zeroExtend wOut (Nat.le_add_right _ _)
  let mut ret : BitVec wOut := 0
  -- fold over the bits of b starting at MSB
  for i in List.range v |>.reverse do
    let_opaque tmp := ret <<< 1
    opaque ret := if b.lsbGet i then polyAdd tmp a else tmp
  return ret

/-- Modulo operation in GF(2)[x] translated from Cryptol reference.

For non-zero b(x), when a(x) = t(x)*b(x) + r(x), mod a b = r.
For zero b(x), we return zero.

```cryptol
pmod : {u, v} (fin u, fin v) => [u] -> [1 + v] -> [v]
pmod x y = if y == 0 then 0/0 else last zs
  where
    degree : [width v]
    degree = last (ds : [2 + v]_)
      where ds = [0/0] # [if yi then i else d | yi <- reverse y | i <- [0..v] | d <- ds ]

    reduce : [1 + v] -> [1 + v]
    reduce u = if u ! degree then u ^ y else u

    powers : [inf][1 + v]
    powers = [reduce 1] # [ reduce (p << 1) | p <- powers ]

    zs = [0] # [ z ^ (if xi then tail p else 0) | xi <- reverse x | p <- powers | z <- zs ]
```
-/
def polyMod (x : BitVec w) (y : BitVec (v+1)) : BitVec v :=
  if y = 0 then 0
  else
    let reduce (a : BitVec (v+1)) : BitVec (v+1) :=
      if a.lsbGet v then polyAdd a y else a
    let_opaque ret : BitVec v := 0
    let_opaque pow : BitVec (v+1) := reduce (BitVec.ofNat (v+1) 1)
    let (ret, _) := List.range w |>.foldl (init := (ret, pow)) fun (ret, pow) i =>
      let_opaque ret := if x.lsbGet i then polyAdd ret (pow.shrink v) else ret
      let_opaque pow := reduce (pow <<< 1)
      (ret, pow)
    ret

section test

-- TODO(WN): We may eventually want to prove these. But to do so, it might make sense
-- to express GF2Poly operations in more abstract terms than the same shifting circuits
-- as given here.
@[reducible]
def is_polyMul_correct (x y : BitVec w) := interp (polyMul x y) = interp x * interp y
@[reducible]
def is_polyMod_correct (x : BitVec w) (y : BitVec (v+1)) := interp (polyMod x y) = interp x % interp y

-- x²+x+1
def a : BitVec 8 := ⟨0b0111, by decide⟩
-- x³+x
def b : BitVec 8 := ⟨0b1010, by decide⟩
-- x⁴+1
def c : BitVec 8 := ⟨0b10001, by decide⟩
-- x³+1
def d : BitVec 8 := ⟨0b01001, by decide⟩
-- x⁷+x+1
def e : BitVec 8 := ⟨0b10000011, by decide⟩

#eval is_polyMul_correct a b
#eval is_polyMul_correct c d
#eval is_polyMod_correct (polyMul a b) e
#eval is_polyMod_correct (polyMul c d) e

end test

end GF2BVPoly

/-- Data of a finite field of the form GF(2^k).

Inspired by the [implementation](https://github.com/project-everest/hacl-star/blob/0a58b6343c2dac2cd87a17f1ecb8233c3947f856/specs/Spec.GaloisField.fst)
in HACL*. -/
structure GaloisField2k where
  k : Nat
  /-- An irreducible polynomial with degree `k`. Representing it explicitly would require `k+1`
  bits, but we know that the highest one is set, and so need only only store the lower `k` bits.
  So the "real" polynomial is `x^k + GF2BVPoly.interp irred`. -/
  irred : BitVec k

/-- Select the smallest integer type able to contain elements of GF(2^k),
or `BitVec` if no integer type can contain them. -/
-- TODO: Specialize the field operations for these.
def GaloisField2k.uint : Nat → Type
  | 0   => UInt8
  | 9   => UInt16
  | 17  => UInt32
  | 33  => UInt64
  | k+1 =>
    if k < 64 then uint k
    else BitVec (k+1)

instance : CoeSort GaloisField2k Type :=
  ⟨fun F => BitVec F.k⟩

namespace GaloisField2k
open GF2BVPoly

variable {F : GaloisField2k}

instance : OfNat F (nat_lit 0) := ⟨BitVec.ofNat F.k 0⟩
instance : OfNat F (nat_lit 1) := ⟨BitVec.ofNat F.k 1⟩

def add (a b : F) : F := polyAdd a b

instance : Add F := ⟨add⟩

def addMany (as : Array F) : F :=
  as.foldl (init := BitVec.ofNat F.k 0) add

-- TODO: use specialized polyMod instead
def irredFull (F : GaloisField2k) :=
  BitVec.ofNat (F.k+1) (1 <<< F.k) ||| F.irred.zeroExtend (F.k+1) (Nat.le_succ _)

def mul (a b : F) : F := polyMod (polyMul a b) (irredFull F)

instance : Mul F := ⟨mul⟩

def pow (k : Nat) (x : F) : F :=
  if hEq : k = 0 then 1
  else
    have : k / 2 < k := Nat.div_lt_self (Nat.zero_lt_of_ne_zero hEq) (by decide)
    if k % 2 = 0 then sq (pow (k / 2) x)
    else mul x (sq (pow (k / 2) x))
where
  sq (x : F) := mul x x

instance : Pow F Nat := ⟨fun x k => pow k x⟩

def inv (x : F) : F :=
  pow (2^F.k - 1) x

end GaloisField2k

open GaloisField2k

def GF256 : GaloisField2k where
  k := 8
  -- x⁸ + x⁴ + x³ + x + 1
  irred := BitVec.ofNat 8 0b11011

-- NOTE(WN): We would like to unroll the loop in `pow` during specialization in order to create
-- an easier SMT problem, but currently WHNFSmt reduction results in some WF internals. It should
-- apply unfolding theorems instead. Z3 can deal with this variant even without unrolling.
/-- `pow2 k = pow (2^k)` -/
def GF256.pow2 (k : Nat) (x : GF256) : GF256 :=
  if k = 0 then x
  else
    let_opaque v : GF256 := pow2 (k-1) x
    mul v v

def GF256.inv' (x : GF256) : GF256 :=
  -- pow 254 x
  let v := pow2 7 x
  let v := mul v (pow2 6 x)
  let v := mul v (pow2 5 x)
  let v := mul v (pow2 4 x)
  let v := mul v (pow2 3 x)
  let v := mul v (pow2 2 x)
  let v := mul v (pow2 1 x)
  v