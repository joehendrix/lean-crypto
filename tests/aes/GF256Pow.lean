import Smt

import Crypto.GF2BV

open GF2BVPoly GaloisField2k

set_option smt.solver.kind "z3" in
example (x : GF256) : GF256.pow2 8 x = x := by
  extract_def GF256.pow2

  extract_def mul
  specialize_def GaloisField2k.mul [GF256] blocking [@polyMod, @polyMul]
  conv at GF256.pow2.def =>
    intro k x
    rw [ GaloisField2k.mul.GF256.specialization ]
  save

  extract_def polyMod
  specialize_def GF2BVPoly.polyMod [16, 8]
  save

  extract_def polyMul
  specialize_def GF2BVPoly.polyMul [8, 8]
  save

  conv at GaloisField2k.mul.GF256.def =>
    intro a b
    rw [ GF2BVPoly.polyMul.«8».«8».specialization,
      GF2BVPoly.polyMod.«16».«8».specialization ]

  smt_show [
    GF256.pow2,
    GaloisField2k.mul.GF256,
    GF2BVPoly.polyMul.«8».«8»,
    GF2BVPoly.polyMod.«16».«8»
  ] 
    
  sorry
