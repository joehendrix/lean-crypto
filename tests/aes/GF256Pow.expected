goal: GF256.pow2 8 x = x

query:
(define-sort Nat () Int)
(define-fun Nat.sub ((x Nat) (y Nat)) Nat (ite (< x y) 0 (- x y)))
(assert (forall ((_uniq.7863 Nat)) (=> (>= _uniq.7863 0) (forall ((_uniq.7864 Nat)) (=> (>= _uniq.7864 0) (>= (Nat.sub _uniq.7863 _uniq.7864) 0))))))
(define-fun |GF2BVPoly.polyMod.«16».«8»| ((x (_ BitVec 16)) (y (_ BitVec 9))) (_ BitVec 8) (let ((ret #b00000000)) (let ((pow (ite (= (not (= ((_ extract 8 8) #b000000001) #b0)) true) (bvxor #b000000001 y) #b000000001))) (let ((ret_1 (ite (= (not (= ((_ extract 0 0) x) #b0)) true) (bvxor ret (ite (<= (+ 0 1) 8) ((_ extract 7 0) pow) #b00000000)) ret))) (let ((pow_1 (ite (= (not (= ((_ extract 8 8) (bvshl pow #b000000001)) #b0)) true) (bvxor (bvshl pow #b000000001) y) (bvshl pow #b000000001)))) (let ((ret_2 (ite (= (not (= ((_ extract 1 1) x) #b0)) true) (bvxor ret_1 (ite (<= (+ 0 1) 8) ((_ extract 7 0) pow_1) #b00000000)) ret_1))) (let ((pow_2 (ite (= (not (= ((_ extract 8 8) (bvshl pow_1 #b000000001)) #b0)) true) (bvxor (bvshl pow_1 #b000000001) y) (bvshl pow_1 #b000000001)))) (let ((ret_3 (ite (= (not (= ((_ extract 2 2) x) #b0)) true) (bvxor ret_2 (ite (<= (+ 0 1) 8) ((_ extract 7 0) pow_2) #b00000000)) ret_2))) (let ((pow_3 (ite (= (not (= ((_ extract 8 8) (bvshl pow_2 #b000000001)) #b0)) true) (bvxor (bvshl pow_2 #b000000001) y) (bvshl pow_2 #b000000001)))) (let ((ret_4 (ite (= (not (= ((_ extract 3 3) x) #b0)) true) (bvxor ret_3 (ite (<= (+ 0 1) 8) ((_ extract 7 0) pow_3) #b00000000)) ret_3))) (let ((pow_4 (ite (= (not (= ((_ extract 8 8) (bvshl pow_3 #b000000001)) #b0)) true) (bvxor (bvshl pow_3 #b000000001) y) (bvshl pow_3 #b000000001)))) (let ((ret_5_1 (ite (= (not (= ((_ extract 4 4) x) #b0)) true) (bvxor ret_4 (ite (<= (+ 0 1) 8) ((_ extract 7 0) pow_4) #b00000000)) ret_4))) (let ((pow_5_1 (ite (= (not (= ((_ extract 8 8) (bvshl pow_4 #b000000001)) #b0)) true) (bvxor (bvshl pow_4 #b000000001) y) (bvshl pow_4 #b000000001)))) (let ((ret_6 (ite (= (not (= ((_ extract 5 5) x) #b0)) true) (bvxor ret_5_1 (ite (<= (+ 0 1) 8) ((_ extract 7 0) pow_5_1) #b00000000)) ret_5_1))) (let ((pow_6 (ite (= (not (= ((_ extract 8 8) (bvshl pow_5_1 #b000000001)) #b0)) true) (bvxor (bvshl pow_5_1 #b000000001) y) (bvshl pow_5_1 #b000000001)))) (let ((ret_6_1 (ite (= (not (= ((_ extract 5 5) x) #b0)) true) (bvxor ret_5_1 (ite (<= (+ 0 1) 8) ((_ extract 7 0) pow_5_1) #b00000000)) ret_5_1))) (let ((pow_6_1 (ite (= (not (= ((_ extract 8 8) (bvshl pow_5_1 #b000000001)) #b0)) true) (bvxor (bvshl pow_5_1 #b000000001) y) (bvshl pow_5_1 #b000000001)))) (let ((ret_7 (ite (= (not (= ((_ extract 6 6) x) #b0)) true) (bvxor ret_6 (ite (<= (+ 0 1) 8) ((_ extract 7 0) pow_6) #b00000000)) ret_6))) (let ((pow_7 (ite (= (not (= ((_ extract 8 8) (bvshl pow_6 #b000000001)) #b0)) true) (bvxor (bvshl pow_6 #b000000001) y) (bvshl pow_6 #b000000001)))) (let ((ret_7_1 (ite (= (not (= ((_ extract 6 6) x) #b0)) true) (bvxor ret_6_1 (ite (<= (+ 0 1) 8) ((_ extract 7 0) pow_6_1) #b00000000)) ret_6_1))) (let ((pow_7_1 (ite (= (not (= ((_ extract 8 8) (bvshl pow_6_1 #b000000001)) #b0)) true) (bvxor (bvshl pow_6_1 #b000000001) y) (bvshl pow_6_1 #b000000001)))) (let ((ret_8 (ite (= (not (= ((_ extract 7 7) x) #b0)) true) (bvxor ret_7 (ite (<= (+ 0 1) 8) ((_ extract 7 0) pow_7) #b00000000)) ret_7))) (let ((pow_8 (ite (= (not (= ((_ extract 8 8) (bvshl pow_7 #b000000001)) #b0)) true) (bvxor (bvshl pow_7 #b000000001) y) (bvshl pow_7 #b000000001)))) (let ((ret_8_1 (ite (= (not (= ((_ extract 7 7) x) #b0)) true) (bvxor ret_7_1 (ite (<= (+ 0 1) 8) ((_ extract 7 0) pow_7_1) #b00000000)) ret_7_1))) (let ((pow_8_1 (ite (= (not (= ((_ extract 8 8) (bvshl pow_7_1 #b000000001)) #b0)) true) (bvxor (bvshl pow_7_1 #b000000001) y) (bvshl pow_7_1 #b000000001)))) (let ((ret_9 (ite (= (not (= ((_ extract 8 8) x) #b0)) true) (bvxor ret_8 (ite (<= (+ 0 1) 8) ((_ extract 7 0) pow_8) #b00000000)) ret_8))) (let ((pow_9 (ite (= (not (= ((_ extract 8 8) (bvshl pow_8 #b000000001)) #b0)) true) (bvxor (bvshl pow_8 #b000000001) y) (bvshl pow_8 #b000000001)))) (let ((ret_9_1 (ite (= (not (= ((_ extract 8 8) x) #b0)) true) (bvxor ret_8_1 (ite (<= (+ 0 1) 8) ((_ extract 7 0) pow_8_1) #b00000000)) ret_8_1))) (let ((pow_9_1 (ite (= (not (= ((_ extract 8 8) (bvshl pow_8_1 #b000000001)) #b0)) true) (bvxor (bvshl pow_8_1 #b000000001) y) (bvshl pow_8_1 #b000000001)))) (let ((ret_5 (ite (= (not (= ((_ extract 9 9) x) #b0)) true) (bvxor ret_9_1 (ite (<= (+ 0 1) 8) ((_ extract 7 0) pow_9_1) #b00000000)) ret_9_1))) (let ((pow_5 (ite (= (not (= ((_ extract 8 8) (bvshl pow_9_1 #b000000001)) #b0)) true) (bvxor (bvshl pow_9_1 #b000000001) y) (bvshl pow_9_1 #b000000001)))) (let ((ret_10 (ite (= (not (= ((_ extract 10 10) x) #b0)) true) (bvxor ret_5 (ite (<= (+ 0 1) 8) ((_ extract 7 0) pow_5) #b00000000)) ret_5))) (let ((pow_10 (ite (= (not (= ((_ extract 8 8) (bvshl pow_5 #b000000001)) #b0)) true) (bvxor (bvshl pow_5 #b000000001) y) (bvshl pow_5 #b000000001)))) (let ((ret_11 (ite (= (not (= ((_ extract 11 11) x) #b0)) true) (bvxor ret_10 (ite (<= (+ 0 1) 8) ((_ extract 7 0) pow_10) #b00000000)) ret_10))) (let ((pow_11 (ite (= (not (= ((_ extract 8 8) (bvshl pow_10 #b000000001)) #b0)) true) (bvxor (bvshl pow_10 #b000000001) y) (bvshl pow_10 #b000000001)))) (let ((ret_12 (ite (= (not (= ((_ extract 12 12) x) #b0)) true) (bvxor ret_11 (ite (<= (+ 0 1) 8) ((_ extract 7 0) pow_11) #b00000000)) ret_11))) (let ((pow_12 (ite (= (not (= ((_ extract 8 8) (bvshl pow_11 #b000000001)) #b0)) true) (bvxor (bvshl pow_11 #b000000001) y) (bvshl pow_11 #b000000001)))) (let ((ret_13 (ite (= (not (= ((_ extract 13 13) x) #b0)) true) (bvxor ret_12 (ite (<= (+ 0 1) 8) ((_ extract 7 0) pow_12) #b00000000)) ret_12))) (let ((pow_13 (ite (= (not (= ((_ extract 8 8) (bvshl pow_12 #b000000001)) #b0)) true) (bvxor (bvshl pow_12 #b000000001) y) (bvshl pow_12 #b000000001)))) (let ((ret_14 (ite (= (not (= ((_ extract 14 14) x) #b0)) true) (bvxor ret_13 (ite (<= (+ 0 1) 8) ((_ extract 7 0) pow_13) #b00000000)) ret_13))) (let ((pow_14 (ite (= (not (= ((_ extract 8 8) (bvshl pow_13 #b000000001)) #b0)) true) (bvxor (bvshl pow_13 #b000000001) y) (bvshl pow_13 #b000000001)))) (let ((ret_15 (ite (= (not (= ((_ extract 15 15) x) #b0)) true) (bvxor ret_14 (ite (<= (+ 0 1) 8) ((_ extract 7 0) pow_14) #b00000000)) ret_14))) (let ((pow_15 (ite (= (not (= ((_ extract 8 8) (bvshl pow_14 #b000000001)) #b0)) true) (bvxor (bvshl pow_14 #b000000001) y) (bvshl pow_14 #b000000001)))) (ite (= y #b000000000) #b00000000 ret_15))))))))))))))))))))))))))))))))))))))))))))
(define-fun |GF2BVPoly.polyMul.«8».«8»| ((a (_ BitVec 8)) (b (_ BitVec 8))) (_ BitVec 16) (let ((v (let ((v_1 (bvshl #b0000000000000000 #b0000000000000001))) (let ((v_1_1 (concat #b00000000 a))) (let ((v_1_1_1 (bvshl #b0000000000000000 #b0000000000000001))) (ite (= (not (= ((_ extract 7 7) b) #b0)) true) (bvxor v_1_1_1 v_1_1) v_1)))))) (let ((v_1 (let ((v_1_1 (bvshl v #b0000000000000001))) (let ((v_1_1_1 (concat #b00000000 a))) (let ((v_1_1_1_1 (bvshl v #b0000000000000001))) (ite (= (not (= ((_ extract 6 6) b) #b0)) true) (bvxor v_1_1_1_1 v_1_1_1) v_1_1)))))) (let ((v_2 (let ((v_2_1 (bvshl v_1 #b0000000000000001))) (let ((v_2_1_1 (concat #b00000000 a))) (let ((v_2_1_1_1 (bvshl v_1 #b0000000000000001))) (ite (= (not (= ((_ extract 5 5) b) #b0)) true) (bvxor v_2_1_1_1 v_2_1_1) v_2_1)))))) (let ((v_3 (let ((v_3_1 (bvshl v_2 #b0000000000000001))) (let ((v_3_1_1 (concat #b00000000 a))) (let ((v_3_1_1_1 (bvshl v_2 #b0000000000000001))) (ite (= (not (= ((_ extract 4 4) b) #b0)) true) (bvxor v_3_1_1_1 v_3_1_1) v_3_1)))))) (let ((v_4 (let ((v_4_1 (bvshl v_3 #b0000000000000001))) (let ((v_4_1_1 (concat #b00000000 a))) (let ((v_4_1_1_1 (bvshl v_3 #b0000000000000001))) (ite (= (not (= ((_ extract 3 3) b) #b0)) true) (bvxor v_4_1_1_1 v_4_1_1) v_4_1)))))) (let ((v_5 (let ((v_5_1 (bvshl v_4 #b0000000000000001))) (let ((v_5_1_1 (concat #b00000000 a))) (let ((v_5_1_1_1 (bvshl v_4 #b0000000000000001))) (ite (= (not (= ((_ extract 2 2) b) #b0)) true) (bvxor v_5_1_1_1 v_5_1_1) v_5_1)))))) (let ((v_6 (let ((v_6_1 (bvshl v_5 #b0000000000000001))) (let ((v_6_1_1 (concat #b00000000 a))) (let ((v_6_1_1_1 (bvshl v_5 #b0000000000000001))) (ite (= (not (= ((_ extract 1 1) b) #b0)) true) (bvxor v_6_1_1_1 v_6_1_1) v_6_1)))))) (let ((v_7 (let ((v_7_1 (bvshl v_6 #b0000000000000001))) (let ((v_7_1_1 (concat #b00000000 a))) (let ((v_7_1_1_1 (bvshl v_6 #b0000000000000001))) (ite (= (not (= ((_ extract 0 0) b) #b0)) true) (bvxor v_7_1_1_1 v_7_1_1) v_7_1)))))) v_7)))))))))
(define-fun GaloisField2k.mul.GF256 ((a (_ BitVec 8)) (b (_ BitVec 8))) (_ BitVec 8) (|GF2BVPoly.polyMod.«16».«8»| (|GF2BVPoly.polyMul.«8».«8»| a b) (bvor #b100000000 (concat #b0 #b00011011))))
(define-fun-rec GF256.pow2 ((k Nat) (x (_ BitVec 8))) (_ BitVec 8) (ite (= k 0) x (GaloisField2k.mul.GF256 (GF256.pow2 (Nat.sub k 1) x) (GF256.pow2 (Nat.sub k 1) x))))
(declare-const x (_ BitVec 8))
(assert (not (= (GF256.pow2 8 x) x)))
(check-sat)
tests/aes/GF256Pow.lean:8:0: warning: declaration uses 'sorry'
